Benchmark cleanup and new and generic timing API

nashif opened the pull request:
Cleanup benchmark and remove code duplication for acquiring timestamps from different source. Make the timing functions generally available and overridable by SoC or boards. The APIs should serve other time sensitive features such as per thread accounting and tracing.

Remove EXECUTION_BENCHMARKING option which was dedicated for single test and polluted all arch and kernel code measuring data points that do not reflect reality. 
For low lever time measurements and tracing we will be using the tracing feature which happens to hit exactly the same spots.

This cleans up the code considerable.

dcpleung committed:
tests: timing_info: remove timing info

Remove this benchmark which was relying on custom tracing points in the
code and was not scalable. Use latency_measure benchmark instead which
is more realistic and measures similar metrics in a fully reproducible
manner and on all supported architectures.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
x86: use LFENCE instead of CPUID before reading TSC for x86_64

According to Intel 64 and IA-32 Architectures Software
Developer’s Manual, volume 3, chapter 8.2.5, LFENCE provides
a more efficient method of controlling memory ordering than
the CPUID instruction. So use LFENCE here, as all 64-bit
CPUs have LFENCE.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
x86: use =A as output for RDTSC on x86-32

The timing_info benchmark on qemu_x86 shows this is a bit faster.

Before:
  START - Time Measurement
  Timing results: Clock frequency: 1000 MHz
  Context switch                               : 896 cycles ,   895 ns
  Interrupt latency                            : 768 cycles ,   767 ns
  Tick overhead                                :14912 cycles , 14911 ns
  Thread creation                              :18688 cycles , 18687 ns
  Thread abort (non-running)                   :49216 cycles , 49215 ns
  Thread abort (_current)                      :55616 cycles , 55615 ns
  Thread suspend                               :11072 cycles , 11071 ns
  Thread resume                                :10272 cycles , 10271 ns
  Thread yield                                 :12213 cycles , 12212 ns
  Thread sleep                                 :17984 cycles , 17983 ns
  Heap malloc                                  :21702 cycles , 21701 ns
  Heap free                                    :15176 cycles , 15175 ns
  Semaphore take with context switch           :19168 cycles , 19167 ns
  Semaphore give with context switch           :18400 cycles , 18399 ns
  Semaphore take without context switch        :2208 cycles ,  2207 ns
  Semaphore give without context switch        :4704 cycles ,  4703 ns
  Mutex lock                                   :1952 cycles ,  1951 ns
  Mutex unlock                                 :7936 cycles ,  7935 ns
  Message queue put with context switch        :20320 cycles , 20319 ns
  Message queue put without context switch     :5792 cycles ,  5791 ns
  Message queue get with context switch        :22112 cycles , 22111 ns
  Message queue get without context switch     :5312 cycles ,  5311 ns
  Mailbox synchronous put                      :27936 cycles , 27935 ns
  Mailbox synchronous get                      :23392 cycles , 23391 ns
  Mailbox asynchronous put                     :11808 cycles , 11807 ns
  Mailbox get without context switch           :20416 cycles , 20415 ns
  Drop to user mode                            :643712 cycles , 643711 ns
  User thread creation                         :652096 cycles , 652095 ns
  Syscall overhead                             :2720 cycles ,  2719 ns
  Validation overhead k_object init            :4256 cycles ,  4255 ns
  Validation overhead k_object permission      :4224 cycles ,  4223 ns
  Time Measurement finished

After:
  START - Time Measurement
  Timing results: Clock frequency: 1000 MHz
  Context switch                               : 896 cycles ,   895 ns
  Interrupt latency                            : 768 cycles ,   767 ns
  Tick overhead                                :14752 cycles , 14751 ns
  Thread creation                              :18464 cycles , 18463 ns
  Thread abort (non-running)                   :48992 cycles , 48991 ns
  Thread abort (_current)                      :55552 cycles , 55551 ns
  Thread suspend                               :10848 cycles , 10847 ns
  Thread resume                                :10048 cycles , 10047 ns
  Thread yield                                 :12213 cycles , 12212 ns
  Thread sleep                                 :17984 cycles , 17983 ns
  Heap malloc                                  :21702 cycles , 21701 ns
  Heap free                                    :15176 cycles , 15175 ns
  Semaphore take with context switch           :19104 cycles , 19103 ns
  Semaphore give with context switch           :18368 cycles , 18367 ns
  Semaphore take without context switch        :1984 cycles ,  1983 ns
  Semaphore give without context switch        :4480 cycles ,  4479 ns
  Mutex lock                                   :1728 cycles ,  1727 ns
  Mutex unlock                                 :7712 cycles ,  7711 ns
  Message queue put with context switch        :20224 cycles , 20223 ns
  Message queue put without context switch     :5568 cycles ,  5567 ns
  Message queue get with context switch        :22016 cycles , 22015 ns
  Message queue get without context switch     :5088 cycles ,  5087 ns
  Mailbox synchronous put                      :27840 cycles , 27839 ns
  Mailbox synchronous get                      :23296 cycles , 23295 ns
  Mailbox asynchronous put                     :11584 cycles , 11583 ns
  Mailbox get without context switch           :20192 cycles , 20191 ns
  Drop to user mode                            :643616 cycles , 643615 ns
  User thread creation                         :651872 cycles , 651871 ns
  Syscall overhead                             :2464 cycles ,  2463 ns
  Validation overhead k_object init            :4032 cycles ,  4031 ns
  Validation overhead k_object permission      :4000 cycles ,  3999 ns
  Time Measurement finished

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
timing: introduce timing functions as a generic feature

Add timing functions and APIs.  This is now used with some of the tests
we have for performance and metrics and will be used whereever timing
informations are needed, for example for tracing, profiling and other
operations where timing info is critical.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
timing: add support for nordic SoCs with RTC timer

Add abstraction for nordic SoCs using Nordic RTC as the source for
timestamps and cycles.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: latency_measure: Using timing functions

Use new timing API instead if local macros and functions. Add new
becnhmarks for threads and semaphore and change the output to be
parseable.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: timestamp: minor cleanup

Minor cleanup and fixes.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
timing: add support for x86

Add initial support for X86 and get timestamps from tsc.

Authored-by: Daniel Leung <daniel.leung@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: default timings for all architectures

Use default if architecture does not have a custom timing
implementation.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: nios2: add timing implementation

Add timing implementation for NIOS2 architecture.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
soc: mec1501: add timing support

Use custom timing implementation specific for this SoC.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

andrewboie commented on arch/arm/core/aarch32/timing.c:
```diff
@@ -0,0 +1,54 @@
+
+uint64_t __weak timing_cycles_to_ns_avg(uint64_t cycles, uint32_t count)
+{
+	return (uint32_t)timing_cycles_to_ns(cycles) / count;
```
would it make more sense to up-cast count to `uint64_t` here?

andrewboie commented on arch/arm/core/aarch32/timing.c:
```diff
@@ -0,0 +1,54 @@
+
+uint64_t __weak timing_cycles_to_ns(uint64_t cycles)
+{
+	return (uint32_t)k_cyc_to_ns_floor64(cycles);
```
why `uint32_t` cast?

andrewboie commented on arch/common/timing.c:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
```
Getting deja vu reading this code.
Then I realized we already have it in arch/arm.
Wouldn't it be better to just keep this file, and implemnnent any differences in the arch layer?

dcpleung replied:
The idea is that this file is for architectures and SoCs without high resolution timers. So just get the values from the default timer driver. Arch code can override this, as in ARM (SysTick) and x86 (TSC). SoC code can also override these for their specific timers, e.g. NRF and MEC1501. This layering is a bit tricky. I am not sure if `__weak` alone can support this.

abrodkin replied:
I might be nitpicking as on ARC default Timers (either ARC Timer0/1 or GFRC in SMP system) is synchronized with CPU clock which makes those timers pretty high-resolution (as least for measurements sake as we don't care about latency of 0.5 a cycle I guess). Or is there any benefit of having architecture- or SoC-specific implementation of these functions other than higher precision of the timer being used?

nashif replied:
@abrodkin  we will need an implementation of those functions for ARC based on the high res timer. Is this the same thing as in arch/arc/core/timestamp.c?

ruuddw replied:
Both arch/arc/core/timestamp.c and the timer driver arcv2_timer0.c use the timer0 that is part of the ARC core and runs at CPU clock frequency. Timestamp.c doesn't seem to take into account timer setting and wrapping in the timer driver, that is likely a problem.
Depending on ARC configuration, there can be a second timer, and 2 separate secure mode timers. There can be a counter as well, and for SMP a global counter.
Problem is that not all of these have to be present in in all ARC instances, ARC users choose a configuration. So we try to work with the minimum, though using more timers/counters would make implementations simpler.

andrewboie commented on arch/nios2/core/timing.c:
```diff
@@ -0,0 +1,65 @@
+		 << 16) |                                                      \
+		((uint32_t)IORD_ALTERA_AVALON_TIMER_SNAPL(TIMER_0_BASE))))
+
+void __weak timing_init(void)
```
what's the use-case for overriding these at the arch instead of common level?

nashif replied:
an architecture can either use the default and not provide its own or implement its own timing functions in a way that can be overridden by an SoC or a board or an application.

andrewboie replied:
I think I see. I'm not sure if `__weak` is the best way to approach this as it's not completely clear to me if this will work with multiple layers of overrides. We might have to use a vtable or something instead even though it's a bit of extra overhead if we want it to be completely flexible.

andrewboie commented on arch/x86/core/ia32/intstub.S:
```diff
@@ -80,13 +84,13 @@ SECTION_FUNC(TEXT, _interrupt_enter)
+	push	%eax
+	push	%edx
+
+	call	read_timer_start_of_isr
```
this is wrong, you have to push ALL caller-saved registers if you are calling into C code since we are at the beginning of the ISR code where nothing has been preserved yet.
the old code pushed only eax/edx because of how 'rdtsc' works.

andrewboie commented on arch/x86/core/ia32/swap.S:
```diff
@@ -65,27 +69,15 @@
-	mov %edx,arch_timing_swap_start+4
-	pop %edx
-	pop %eax
+	push	%eax
```
I know the old code was being used as a template, but it was written by someone who wasn't the sharpest tool in the shed.

the 32-bit SYS V ABI passes in arguments via the stack, so unless I'm missing something you don't need to preserve either eax or edx as neither of these caller-saved registers are used here

andrewboie commented on arch/x86/core/ia32/userspace.S:
```diff
@@ -375,16 +379,13 @@ SECTION_FUNC(TEXT, z_x86_userspace_enter)
-	mov %edx,arch_timing_enter_user_mode_end+4
-	pop %edx
-	pop %eax
+	push	%eax
```
i don't think these registers need to be preserved since right after this we just `iret`

andrewboie commented on arch/x86/core/intel64/locore.S:
```diff
@@ -288,7 +286,13 @@ mxcsr:	.long X86_MXCSR_SANE
 z_x86_switch:
 #ifdef CONFIG_EXECUTION_BENCHMARKING
-	read_tsc arch_timing_swap_start
+	push	%rax
```
these regs are not in use

andrewboie commented on arch/x86/core/intel64/userspace.S:
```diff
@@ -326,16 +331,13 @@ z_x86_userspace_enter:
-	mov %edx,arch_timing_enter_user_mode_end+4
-	pop %rdx
-	pop %rax
+	push	%rax
```
I don't think RAX needs preservation, note the 'movq' right below this.
RDX needs to be saved, it has the 'p2' argument

andrewboie commented on arch/x86/core/intel64/locore.S:
```diff
@@ -593,7 +597,13 @@ EXCEPT(Z_X86_OOPS_VECTOR);
 irq:
 #ifdef CONFIG_EXECUTION_BENCHMARKING
-	read_tsc arch_timing_irq_start
+	push	%rax
```
All caller-saved regs must be saved before you call into C code here

andrewboie commented on arch/x86/timing.c:
```diff
@@ -0,0 +1,85 @@
+	uint32_t cyc_start = k_cycle_get_32();
+	uint64_t tsc_start = z_tsc_read();
+
+	k_busy_wait(10 * USEC_PER_MSEC);
```
seems a fairly short timing loop, does this produce stable/accurate values on, say, and upsquared?

dcpleung replied:
Seems stable to me. The differences between different runs are in the kHz level.

tyrantlizard77 reviewed (commented):
the asm code for x86 has the odor of copypasta and needs more detailed attention for each and every place an instrumentation point occurs

andrewboie commented:
Please, at least when testing this, hack qemu_x86 and qemu_x86_64 or create new targets to globally enable these instrumentation points to help catch issues with the assembly code changes

dcpleung commented:
The `timing_info` benchmark is using these functions so code is actually being used.

dcpleung reviewed (commented):
About the x86 assembly, they were being tesed with the benchmark apps. The push/pop pairs were added until the apps ran without crashing (e.g. push/pop of `rcx` in only one place but not others). We can add the other registers in there for sake of completion, if you insist.

tyrantlizard77 reviewed (dismissed):
>  tesed with the benchmark apps. The push/pop pairs were added until the apps ran without crashing

All this says is you were getting lucky with the particular benchmark apps tested, or what the registers the compiler decided to use in the current implementation of the C functions and optimization level.

> We can add the other registers in there for sake of completion, if you insist.

You have to consider the register environment of each call site when you are calling C out of assembly code.
At interrupt entry, nothing has been saved. The registers are just whatever was present for the interrupted process. The ABI specifies that when you call into C code, the caller-saved registers are considered volatile and will be discarded, and it is fair game for the C code to do what it likes with them. This will corrupt the interrupted process.

Some CPUs will save register banks when taking interrupts or exceptions, but on x86 this doesn't happen. As these instrumentation points are added to the ASM code in other arches, it's critical to understand both the calling convention ABI and what the CPU does and does not do for you. x86 doesn't save anything in HW which is why the IRQ and exception entry and exit code is like it is.

This wiki page refers to these as "scratch" registers, "volatile" and "caller-saved" are synonyms.
https://wiki.osdev.org/System_V_ABI
https://en.wikipedia.org/wiki/X86_calling_conventions#Caller-saved_(volatile)_registers
https://en.wikipedia.org/wiki/X86_calling_conventions#System_V_AMD64_ABI

Now in other spots in the ASM code, these caller-saved registers may be currently unused at the time the C code is called, in which case pushing/popping them is unnecessary overhead.

This instrumentation framework has to work for general-purpose applications. We must be rigorous about register state. I'm not being pedantic "for the sake of completion" here we really must do this.

nashif commented:
ok, was planning to get rid of the EXECUTION_BENCHMARKING mess in a later PR but I guess we can do it here and avoid having to change the code that will be dropped later anyways. Rationale for removing this:

```
    This code had one purpose only, feed timing information into a test and
    was not used by anything else. The custom trace points unfortunatly were
    not accurate and this test was delivering informatin that conflicted
    with other tests we have due to placement of such trace points in the
    architecture and kernel code.

    For such measurements we are planning to use the tracing functionality
    in a special mode that would be used for metrics without polluting the
    architecture and kernel code with additional tracing and timing code.

    Furthermore, much of the assembly code used had issues
```

We are now using latency_measure as a one stop test for measuring the metrics and remove the duplication and confusing and conflicting results both testing were providing.

nashif committed:
doc: remove benchmarking section from docs

This section does not exist anymore.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

abrodkin reviewed (commented):
That's a good one especially considering my earlier attempt to improve on one separate architecture here https://github.com/zephyrproject-rtos/zephyr/pull/18751.

Though what's now missing here (after removal of `timing_info`) is IRQ latency measurement which I guess hard to do in a completely generic manner. I mean this one:
```
2. Interrupt latency	
   Time taken from the start of the common interrupt handler till the actual ISR handler being called.
```
And back in the day we had a lengthly discussion on how to do it right and how important that might be.
Maybe @ruuddw has something to add on that as well?

ruuddw commented:
Getting rid of EXECUTION_BENCHMARKING cleans up indeed.I don't think tracing covers all benchmark points yet, and not sure it will have the same (low) overhead. But for minimal overhead arch specific instrumentation and tricks would be needed: can be done ad-hoc for the use-cases where required (I expect mostly during development/optimization work). For other use-cases the tracing approach seem fine to me (e.g. performance regression testing).

scottwcpg reviewed (approved):
MCHP SoC hi-res. timer code looks OK.  NOTE: The basic timer will lose its clock in deep sleep. Users of the API should be aware it can't be used across deep sleep.

ioannisg commented on soc/arm/nordic_nrf/timing.c:
```diff
@@ -0,0 +1,68 @@
+#if defined(CONFIG_NRF_RTC_TIMER)
+
+
+#define CYCLES_PER_SEC (16000000 / (1 << NRF_TIMER2->PRESCALER))
```
I think it's good to have a build-time assert here, since it is possible in the future to have nRF SoCs without a 3rd timer instance (TIMER2)

nashif replied:
right, that why the ifdef above, but for other time sources we will have other implementation of those functions.

ioannisg replied:
> > @nashif @dcpleung if you have this board could you investigate?
> 
> This issue existed in master already with nrf51 with latency_measure and other test using the original functions on top of RTC timer. I am not sure how I can fix those, so if you a suggestion, please go ahead, but I do not think they should block this PR. We could disable timeing api for this SoC until we have an appropriate implementation that works.

I don't want to block this PR, but you should at least address the second point i made above (i.e. how to calculate the averages).

nashif commented on soc/arm/nordic_nrf/CMakeLists.txt:
```diff
@@ -6,3 +6,5 @@ zephyr_sources(
   validate_enabled_instances.c
   )
+
+zephyr_library_sources_ifdef(CONFIG_TIMING_FUNCTIONS timing.c)
```
right now this is a catchall for all nrf, if some SoCs have a different source, they will need to provide own implementation.

ioannisg replied:
As I mentioned, Cortex-M SoCs (including some nRF) have DWT which could be used for calculations. So we would end up with two timing.c source files; we need to be able to select one of them. So this is OK for now, but might need to change.

nashif commented on soc/arm/nordic_nrf/timing.c:
```diff
@@ -0,0 +1,68 @@
+#include <timing/timing.h>
+#include <nrfx.h>
+
+#if defined(CONFIG_NRF_RTC_TIMER)
```
in case you have some other timer, given that this file covers everything under nordic_nrf.. If this is not going to be the case, it can be removed.

ioannisg replied:
No, i think this makes no sense.

Being able to use the TIMER2 peripheral for timing measurements has nothing to do with whether NRF_RTC is used for timing. (You could use the SysTick for system timer and still be able to use TIMER2 for timing calculations in timing.c)

Corner case: qemu_cortex_m0 which uses the NRF_TIMER0 for system timer. But still TIMER2 is available for you here. So the #if defined(CONFIG_NRF_RTC_TIMER) is not useful and should be removed.

ioannisg replied:
@nashif if you remove this #if defined(CONFIG_NRF_RTC_TIMER)
the example will compile for qemu_cortex_m0.

ioannisg committed:
arch: arm: cortex_m: conditionally select ARCH_HAS_TIMING_FUNCTIONS

Cortex-M SoCs implement (optionally) the Data Watchpoint and
Tracing Unit (DWT), which can be used for timing functions.
Select the corresponding ARCH capability if the SoC implements
the DWT.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: cortex-m: implement timing.c based on DWT

For Cortex-M platforms with DWT we implement
the timing API (timing.c).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

andrewboie commented on arch/x86/core/ia32/swap.S:
```diff
@@ -65,30 +65,7 @@
-
-
 #if defined(CONFIG_TRACING)
 	pushl	%eax
```
I think the push/pop of EAX can be dropped here, sys_trace_thread_switched_out takes no arguments and we're not using it in this ASM code at the time it is called.

tyrantlizard77 reviewed (commented):
Looking much better, thanks!

Noted one spot where I think we're pushing/popping EAX unnecessarily.

Not loving the use of `__weak` for multiple levels of overrides of timing functions. The one that gets used is (IIRC) the last one that the linker finds, which may or may not follow a kernel -> arch -> soc -> board hierarchy unless the makefiles are set up just right. This might be something where we could get bitten unless we're very careful. Is there another way?

ioannisg commented:
I get this on nRF51, which does not seem right:

```
*** Booting Zephyr OS build zephyr-v2.3.0-2343-g2eb1c191cdba  ***
START - Time Measurement
Timing results: Clock frequency: 16 MHz
Average thread context switch using yield                   : 2148553 cycles , 134284608 ns
Average context switch time between threads (coop)          :       0 cycles ,       14 ns
Switch from ISR back to interrupted thread                  :     612 cycles ,    38250 ns
Time from ISR to executing a different thread               :     525 cycles ,    32812 ns
Time to create a thread (without start)                     :     287 cycles ,    17937 ns
Time to start a thread                                      :     445 cycles ,    27812 ns
Time to suspend a thread                                    :     508 cycles ,    31750 ns
Time to resume a thread                                     :     538 cycles ,    33625 ns
Time to abort a thread (not running)                        :     403 cycles ,    25187 ns
Average semaphore signal time                               : 4294945 cycles , 268434109 ns
Average semaphore test time                                 :       6 cycles ,      408 ns
Semaphore take time (context switch)                        :     589 cycles ,    36812 ns
Semaphore give time (context switch)                        :     690 cycles ,    43125 ns
Average time to lock a mutex                                :      36 cycles ,     2282 ns
Average time to unlock a mutex                              :      29 cycles ,     1843 ns
===================================================================
PROJECT EXECUTION SUCCESSFUL

```

@nashif @dcpleung if you have this board could you investigate?

ioannisg reviewed (changes requested):
@nashif I see two issues with the code here:
1. the nRF timing.c does not work as-is for nRF51 because the nrF51 TIMER2 does not have a full-range bit-mode (only 8-bit and 16-bit modes are supported)
   - You could of course switch to using TIMER0 in the timing.c, but this may interfere with other thing on-board that use TIMER0, e.g. BLE controller). If we believe that TIMING_FUNCTIONS should be used for basic benchmrking, without enabling complext subsystems, that's fine to use this option.
   - Otherwise, you can keep TIMER2 but you need to do some little math around the calculations.

2. The second problem I see is that the average calculations are performed by
   - taking an initial timestamp,
   - running X number of iterations
   - taking an end timestamp
Any counter overflows will corrupt your measurements. And for example in nRF51, 16-bit counter will overflow quickly. I believe that we can fix this by storing the timestamps at every iteration and doing the
averaging at the end of the loop (or stgh similar)

nashif commented:
> @nashif @dcpleung if you have this board could you investigate?

This issue existed in master already with nrf51 with latency_measure and other test using the original functions on top of RTC timer. I am not sure how I can fix those, so if you a suggestion, please go ahead, but I do not think they should block this PR. We could disable timeing api for this SoC until we have an appropriate implementation that works.

nashif commented:
>     * You could of course switch to using TIMER0 in the timing.c, but this may interfere with other thing on-board that use TIMER0, e.g. BLE controller). If we believe that TIMING_FUNCTIONS should be used for basic benchmrking, without enabling complext subsystems, that's fine to use this option.

This is not going to be for benchmarking only, so we should avoid conflicts.

andrewboie commented:
removed asm looks OK.

this one critical for 2.4? not quite convinced use of __weak in this way is the best idea for inheritance, we could get unpleasant surprises if the linker encounters the implementations in an order that doesn't correspond to the intended inheritance hierarchy

nashif commented:
> this one critical for 2.4? not quite convinced use of __weak in this way is the best idea for inheritance, we could get unpleasant surprises if the linker encounters the implementations in an order that doesn't correspond to the intended inheritance hierarchy

Can't think of a better way to allow architectures variants, SoCs and boards to override a default implementation. We already do this for power management.

We need this for 2.4 because we have multiple users depending on this level of abstraction. The way it was implemented (as test code only) and the way we have been sourcing timestamp data for tracing, profiling has been inconsistent and misleading.

ioannisg commented on tests/benchmarks/latency_measure/src/thread_switch_yield.c:
```diff
@@ -79,8 +74,7 @@ void thread_switch_yield(void)
 	/* get the number of cycles it took to do the test */
-	TIMING_INFO_PRE_READ();
-	timestamp_end = TIMING_INFO_OS_GET_TIME();
+	timestamp_end = timing_counter_get();
```
the while loop may take long enough so the underlying cycles' counter infrastructure, which supplies measurements to timing_counter_get() may overflow and the timestamp_end - timestamp_start won't correspond to the elapsed time.

ioannisg reviewed (dismissed):
Still needs to address the issues in the tests, with respect to how average timings are calculated.

ioannisg committed:
soc: arm: nordic nrf: use bitmode-16 for timing measurements in nrf51

nRF51 TIMER2 periperhal does not have the 32-bit
bitmode, so we need to fallback to the bitmode 16.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
soc: arm: nrf: conditionally employ nRF-specific timing framework

Employ the nRF-specific timing calculations framework
(based on TIMER peripheral) only if the DWT is not present
on the SoC.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

andrewboie commented:
> Can't think of a better way to allow architectures variants, SoCs and boards to override a default implementation. We already do this for power management.

Let me ask this: let's consider the most complex case and say we have __weak implementations for some API at the core kernel, arch, soc, and board level.

What is in place to enforce this hierarchy ordering from the linker's perspective? I do not see it.

nashif commented:
> What is in place to enforce this hierarchy ordering from the linker's perspective? I do not see it.

ok, makes sense. Need to think about another approach for this type of implementation. This probably will be the same problem with power management and needs fixing there as well.

andrewboie commented:
> This probably will be the same problem with power management and needs fixing there as well.

I concur, I think if this doesn't maintain the expected ordering hierarchy, very nasty surprises could occur without an intuitive fix.

We might consider a vtable, like our driver model.

nashif committed:
benchmarking: remove execution benchmarking code

This code had one purpose only, feed timing information into a test and
was not used by anything else. The custom trace points unfortunatly were
not accurate and this test was delivering informatin that conflicted
with other tests we have due to placement of such trace points in the
architecture and kernel code.

For such measurements we are planning to use the tracing functionality
in a special mode that would be used for metrics without polluting the
architecture and kernel code with additional tracing and timing code.

Furthermore, much of the assembly code used had issues.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
Signed-off-by: Daniel Leung <daniel.leung@intel.com>

nashif commented:
Using Kconfig for this now, removed __weak usage.

nashif commented:
now also includes #27932

ioannisg commented on soc/arm/nordic_nrf/Kconfig.defconfig:
```diff
@@ -33,6 +33,11 @@ config SYS_POWER_MANAGEMENT
 config BUILD_OUTPUT_HEX
 	default y
 
+if !CORTEX_M_DWT
```
Not sure if this is correct (i.e. we do not need the if !CORTEX_M_DWT:
The nRF SoCs _always have SOC_TIMING_FUNCTIONS and it should be default y.
The CMakeLists.txt , though, can make sure the ARCH (DWT) timing.c will be used, if present. Correct?

ioannisg commented on arch/arm/core/aarch32/cortex_m/CMakeLists.txt:
```diff
@@ -15,6 +15,16 @@ zephyr_library_sources(
 
 zephyr_library_sources_ifdef(CONFIG_DEBUG_COREDUMP coredump.c)
 
+if(CONFIG_ARCH_HAS_TIMING_FUNCTIONS AND
```
I would remove this external if and _prioritize_ the ARCH-solution (DWT) if present, against SoC and Board.

ioannisg commented on arch/arm/core/aarch32/cortex_m/timing.c:
```diff
@@ -0,0 +1,168 @@
+#endif /* CONFIG_SOC_FAMILY_NRF */
+}
+
+void __weak timing_init(void)
```
I guess these __weak attributes need to be removed, now.

ioannisg commented on soc/Kconfig:
```diff
@@ -41,3 +41,9 @@ config SOC_DEPRECATED_RELEASE
+config SOC_HAS_TIMING_FUNCTIONS
+	bool
+	help
+	  Should be selected if board provides custom method for retrieving
```
SoC, not Board. Otherwise it is the same text as BOARD_HAS_TIMING_FUNCTIONS

ioannisg reviewed (commented):
Left some comments that should be fixed, but I think this PR could go in as is, provided the CI passes.
(I.e. the comments may be treated as implementation bugs and could be fixed later)
